File: internal/orchestrator/orchestrator.go
package orchestrator

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/rs/zerolog"
)

// Config holds orchestrator configuration
type Config struct {
	MaxRecursionDepth int
	MaxIterations     int
	CacheEnabled      bool
	CacheTTL          time.Duration
	WorkDir           string
	StateFile         string
}

// DefaultConfig returns default configuration
func DefaultConfig() *Config {
	return &Config{
		MaxRecursionDepth: 10,
		MaxIterations:     1000,
		CacheEnabled:      true,
		CacheTTL:          DefaultCacheTTL,
		WorkDir:           ".",
		StateFile:         StateFileName,
	}
}

// Orchestrator manages the trampoline-based recursion pattern
type Orchestrator struct {
	config      *Config
	logger      zerolog.Logger
	stack       []Task
	currentTask Task
	results     map[string]interface{}
	stats       Stats
	dispatcher  SubagentDispatcher
}

// SubagentDispatcher is a function that dispatches work to a subagent
// Returns either a ContinuationRequest or AnalysisResult
type SubagentDispatcher func(ctx context.Context, task *Task) (*SubagentResult, error)

var (
	ErrMaxDepthExceeded      = errors.New("maximum recursion depth exceeded")
	ErrMaxIterationsExceeded = errors.New("maximum iterations exceeded")
	ErrNoDispatcher          = errors.New("no subagent dispatcher configured")
)

// New creates a new orchestrator
func New(config *Config, logger zerolog.Logger) *Orchestrator {
	if config == nil {
		config = DefaultConfig()
	}

	return &Orchestrator{
		config:  config,
		logger:  logger,
		stack:   make([]Task, 0),
		results: make(map[string]interface{}),
		stats: Stats{
			StartTime: time.Now(),
		},
	}
}

// SetDispatcher sets the subagent dispatcher function
func (o *Orchestrator) SetDispatcher(dispatcher SubagentDispatcher) {
	o.dispatcher = dispatcher
}

// Stats returns the current statistics
func (o *Orchestrator) GetStats() Stats {
	return o.stats
}

// AnalyzeDocument is the main entry point for document analysis
func (o *Orchestrator) AnalyzeDocument(ctx context.Context, documentPath, query string) (*AnalysisResult, error) {
	if o.dispatcher == nil {
		return nil, ErrNoDispatcher
	}

	// Reset stats for new analysis
	o.stats = Stats{
		StartTime: time.Now(),
	}

	// Try to restore state if exists
	if o.HasState() {
		if err := o.LoadState(); err != nil {
			o.logger.Warn().Err(err).Msg("Failed to restore state, starting fresh")
		} else {
			o.logger.Info().Msg("Resumed from previous state")
		}
	}

	// Initialize current task if starting fresh
	if o.currentTask.AgentType == "" {
		o.currentTask = Task{
			AgentType:       "Explorer",
			TaskDescription: query,
			Context: map[string]interface{}{
				"document_path": documentPath,
				"query":         query,
			},
			Depth:        0,
			ChildResults: make(map[string]interface{}),
		}
	}

	// Trampoline loop
	iterations := 0
	for {
		iterations++

		// Safety checks (check depth first for better error messages)
		if o.currentTask.Depth > o.config.MaxRecursionDepth {
			return nil, ErrMaxDepthExceeded
		}

		if iterations > o.config.MaxIterations {
			return nil, ErrMaxIterationsExceeded
		}

		// Track max depth
		if o.currentTask.Depth > o.stats.MaxDepthReached {
			o.stats.MaxDepthReached = o.currentTask.Depth
		}

		o.logger.Info().
			Str("agent", o.currentTask.AgentType).
			Int("depth", o.currentTask.Depth).
			Int("stack_size", len(o.stack)).
			Msg("Dispatching subagent")

		var result *SubagentResult

		// Check cache first
		if cachedResult := o.CheckCache(&o.currentTask); cachedResult != nil {
			o.stats.CacheHits++
			o.logger.Debug().Msg("Using cached result")
			result = &SubagentResult{
				Type:     ResultTypeAnalysis,
				Analysis: cachedResult,
			}
		} else {
			// Dispatch to subagent
			var err error
			result, err = o.dispatcher(ctx, &o.currentTask)
			if err != nil {
				return nil, fmt.Errorf("subagent dispatch failed: %w", err)
			}
			o.stats.TotalSubagentCalls++
		}

		// Process result (both cached and dispatched results)
		if err := o.processResult(ctx, result); err != nil {
			return nil, err
		}

		// Check if we're done
		if len(o.stack) == 0 && result.IsAnalysis() {
			// Clean up state file on completion
			o.ClearState()

			return result.Analysis, nil
		}

		// Save state after each iteration
		if err := o.SaveState(); err != nil {
			o.logger.Warn().Err(err).Msg("Failed to save state")
		}
	}
}

// processResult handles the result from a subagent dispatch
func (o *Orchestrator) processResult(ctx context.Context, result *SubagentResult) error {
	if result.IsContinuation() {
		// CONTINUATION: Push current task to stack, create new task
		o.logger.Debug().
			Str("continuation_agent", result.Continuation.AgentType).
			Str("return_to", result.Continuation.ReturnTo).
			Msg("Continuation requested")

		// Push current task onto stack
		o.stack = append(o.stack, o.currentTask)

		// Create new task from continuation
		returnTo := result.Continuation.ReturnTo
		o.currentTask = Task{
			AgentType:       result.Continuation.AgentType,
			TaskDescription: result.Continuation.Task,
			Context:         result.Continuation.Context,
			Depth:           o.currentTask.Depth + 1,
			ReturnTo:        &returnTo,
			ChildResults:    make(map[string]interface{}),
		}

		return nil // Continue trampolining
	}

	if result.IsAnalysis() {
		// RESULT: Store result and pop stack
		o.logger.Debug().
			Int("token_count", result.Analysis.TokenCount).
			Float64("cost_usd", result.Analysis.CostUSD).
			Msg("Analysis result received")

		// Update stats
		o.stats.TotalTokens += result.Analysis.TokenCount
		o.stats.TotalCostUSD += result.Analysis.CostUSD

		// Store result in cache
		if err := o.StoreCache(&o.currentTask, result.Analysis); err != nil {
			o.logger.Warn().Err(err).Msg("Failed to cache result")
		}

		// If stack is not empty, inject result into parent and continue
		if len(o.stack) > 0 {
			// Store result
			if o.currentTask.ReturnTo != nil {
				o.results[*o.currentTask.ReturnTo] = result.Analysis
			}

			// Pop parent task from stack
			parentTask := o.stack[len(o.stack)-1]
			o.stack = o.stack[:len(o.stack)-1]

			// Inject child results into parent
			if o.currentTask.ReturnTo != nil {
				if parentTask.ChildResults == nil {
					parentTask.ChildResults = make(map[string]interface{})
				}
				parentTask.ChildResults[*o.currentTask.ReturnTo] = result.Analysis
			}

			o.currentTask = parentTask

			o.logger.Debug().
				Int("new_depth", o.currentTask.Depth).
				Int("stack_size", len(o.stack)).
				Msg("Popped stack, continuing with parent task")

			return nil // Continue trampolining
		}

		// Stack is empty and we have a result - we're done
		return nil
	}

	return fmt.Errorf("unknown result type")
}

// PlaceholderDispatcher is a placeholder for the actual subagent dispatcher
// This will be replaced when Claude Code's Agent SDK integration is available
func PlaceholderDispatcher(ctx context.Context, task *Task) (*SubagentResult, error) {
	// Simulate some work
	time.Sleep(100 * time.Millisecond)

	// For now, always return an analysis result (no recursion)
	// In real implementation, Explorer might return ContinuationRequest
	// to spawn Worker subagents for deeper analysis

	return &SubagentResult{
		Type: ResultTypeAnalysis,
		Analysis: &AnalysisResult{
			Type:       "RESULT",
			Content:    fmt.Sprintf("Analysis of '%s' at depth %d", task.TaskDescription, task.Depth),
			Metadata:   task.Context,
			TokenCount: 1000,
			CostUSD:    0.003,
		},
	}, nil
}

// ParseSubagentResponse parses JSON response from subagent
func ParseSubagentResponse(data []byte) (*SubagentResult, error) {
	// Try to determine type by checking for "type" field
	var typeCheck struct {
		Type string `json:"type"`
	}

	if err := json.Unmarshal(data, &typeCheck); err != nil {
		return nil, err
	}

	switch typeCheck.Type {
	case "CONTINUATION":
		var cont ContinuationRequest
		if err := json.Unmarshal(data, &cont); err != nil {
			return nil, err
		}
		return &SubagentResult{
			Type:         ResultTypeContinuation,
			Continuation: &cont,
		}, nil

	case "RESULT":
		var result AnalysisResult
		if err := json.Unmarshal(data, &result); err != nil {
			return nil, err
		}
		return &SubagentResult{
			Type:     ResultTypeAnalysis,
			Analysis: &result,
		}, nil

	default:
		return nil, fmt.Errorf("unknown result type: %s", typeCheck.Type)
	}
}

File: src/orchestrator.py
#!/usr/bin/env python3
"""
RLM Orchestrator - Trampoline-based recursive document analysis

This module implements the core orchestration logic for the RLM plugin,
enabling recursive document analysis through a trampoline pattern that
respects Claude Code's "subagents can't spawn subagents" constraint.
"""
import json
import os
from typing import Dict, Any, Optional, List
from dataclasses import dataclass, asdict
from datetime import datetime
import hashlib

@dataclass
class ContinuationRequest:
    """Request from subagent to recurse deeper"""
    agent_type: str  # "Explorer", "Worker", "Synthesizer"
    task: str
    context: Dict[str, Any]
    return_to: str  # Identifier for result storage
    metadata: Optional[Dict[str, Any]] = None

    def to_dict(self):
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict):
        return cls(**data)

@dataclass
class AnalysisResult:
    """Completed analysis from subagent"""
    content: str
    metadata: Dict[str, Any]
    token_count: int
    cost_usd: float

    def to_dict(self):
        return asdict(self)

@dataclass
class Task:
    """Represents a task for a subagent"""
    agent_type: str
    task_description: str
    context: Dict[str, Any]
    depth: int
    return_to: Optional[str] = None
    child_results: Optional[Dict[str, Any]] = None

    def to_dict(self):
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict):
        return cls(**data)

class RLMOrchestrator:
    """Manages trampoline-based recursive document analysis"""

    def __init__(self, config: Optional[Dict] = None):
        self.config = config or self.load_config()
        self.state_file = ".rlm_state.json"
        self.results_dir = "analysis"
        self.cache_dir = ".rlm_cache"

        # Create directories
        os.makedirs(self.results_dir, exist_ok=True)
        os.makedirs(self.cache_dir, exist_ok=True)

        # Statistics
        self.stats = {
            "total_subagent_calls": 0,
            "total_tokens": 0,
            "total_cost_usd": 0.0,
            "max_depth_reached": 0,
            "cache_hits": 0,
            "start_time": datetime.now().isoformat()
        }

    def load_config(self) -> Dict:
        """Load plugin configuration"""
        config_path = os.path.expanduser("~/.claude-rlm/config.json")
        if os.path.exists(config_path):
            with open(config_path) as f:
                return json.load(f)

        # Default config
        return {
            "max_recursion_depth": 10,
            "cache_enabled": True,
            "cache_ttl_hours": 24,
            "parallel_branches": 3,
            "cost_tracking": True
        }

    def load_state(self) -> Optional[Dict]:
        """Load orchestrator state"""
        if os.path.exists(self.state_file):
            with open(self.state_file) as f:
                data = json.load(f)
                # Convert task dicts back to Task objects
                data['stack'] = [Task.from_dict(t) for t in data['stack']]
                data['current_task'] = Task.from_dict(data['current_task'])
                return data

        return None

    def save_state(self, stack: List[Task], current_task: Task, results: Dict):
        """Persist orchestrator state"""
        state = {
            "stack": [t.to_dict() for t in stack],
            "current_task": current_task.to_dict(),
            "results": results,
            "stats": self.stats,
            "timestamp": datetime.now().isoformat()
        }

        with open(self.state_file, 'w') as f:
            json.dump(state, f, indent=2)

    def clear_state(self):
        """Clear state after completion"""
        if os.path.exists(self.state_file):
            os.remove(self.state_file)

    def get_cache_key(self, task: Task) -> str:
        """Generate cache key for task"""
        key_data = f"{task.agent_type}:{task.task_description}:{json.dumps(task.context, sort_keys=True)}"
        return hashlib.sha256(key_data.encode()).hexdigest()

    def check_cache(self, task: Task) -> Optional[AnalysisResult]:
        """Check if result is cached"""
        if not self.config["cache_enabled"]:
            return None

        cache_key = self.get_cache_key(task)
        cache_path = os.path.join(self.cache_dir, f"{cache_key}.json")

        if os.path.exists(cache_path):
            # Check TTL
            mtime = os.path.getmtime(cache_path)
            age_hours = (datetime.now().timestamp() - mtime) / 3600

            if age_hours < self.config["cache_ttl_hours"]:
                with open(cache_path) as f:
                    data = json.load(f)
                    self.stats["cache_hits"] += 1
                    return AnalysisResult(**data)

        return None

    def save_cache(self, task: Task, result: AnalysisResult):
        """Save result to cache"""
        if not self.config["cache_enabled"]:
            return

        cache_key = self.get_cache_key(task)
        cache_path = os.path.join(self.cache_dir, f"{cache_key}.json")

        with open(cache_path, 'w') as f:
            json.dump(result.to_dict(), f, indent=2)

    def dispatch_subagent(self, task: Task) -> Any:
        """
        Dispatch task to appropriate subagent

        NOTE: This is where Claude Code's native subagent spawning happens.
        The actual implementation will use Claude's Task tool or Agent SDK.
        This is a placeholder showing the interface.
        """
        # Check cache first
        cached = self.check_cache(task)
        if cached:
            print(f"✅ Cache hit for {task.agent_type} task")
            return cached

        # Update stats
        self.stats["total_subagent_calls"] += 1
        self.stats["max_depth_reached"] = max(self.stats["max_depth_reached"], task.depth)

        print(f"\n🎯 Spawning {task.agent_type} subagent (depth {task.depth})")
        print(f"   Task: {task.task_description[:80]}...")

        # Format task for subagent
        subagent_prompt = self.format_task_for_subagent(task)

        # This would be replaced with actual Claude Code subagent spawning:
        # result = spawn_subagent(task.agent_type, subagent_prompt)

        # For now, return placeholder that shows expected structure
        print(f"   [Subagent would execute here]")
        return {
            "type": "RESULT",
            "content": f"Analysis of {task.task_description}",
            "metadata": {"task": task.to_dict()},
            "token_count": 1000,
            "cost_usd": 0.003
        }

    def format_task_for_subagent(self, task: Task) -> str:
        """Format task into subagent prompt"""
        prompt_parts = [
            f"You are a {task.agent_type} subagent in a recursive document analysis system.",
            f"\nTask: {task.task_description}",
            f"\nDepth: {task.depth}/{self.config['max_recursion_depth']}",
        ]

        if task.context:
            prompt_parts.append(f"\nContext:\n{json.dumps(task.context, indent=2)}")

        if task.child_results:
            prompt_parts.append(f"\nChild Results:\n{json.dumps(task.child_results, indent=2)}")

        prompt_parts.append("\n---")
        prompt_parts.append("\nIf you need deeper analysis, return a ContinuationRequest:")
        prompt_parts.append("""
```json
{
  "type": "CONTINUATION",
  "agent_type": "Worker",
  "task": "Detailed description",
  "context": {"key": "value"},
  "return_to": "unique_identifier"
}
```
""")
        prompt_parts.append("\nOtherwise, return your analysis results directly.")

        return "\n".join(prompt_parts)

    def is_continuation(self, result: Any) -> bool:
        """Check if result is a continuation request"""
        if isinstance(result, dict):
            return result.get("type") == "CONTINUATION"
        return False

    def create_task_from_continuation(self, parent_task: Task, cont: Dict) -> Task:
        """Create new task from continuation request"""
        return Task(
            agent_type=cont["agent_type"],
            task_description=cont["task"],
            context=cont["context"],
            depth=parent_task.depth + 1,
            return_to=cont["return_to"]
        )

    def analyze_document(self, document_path: str, query: str) -> AnalysisResult:
        """
        Main entry point for document analysis

        Args:
            document_path: Path to document or directory to analyze
            query: Analysis objective

        Returns:
            Final analysis result
        """
        # Check for existing state (resume capability)
        existing_state = self.load_state()

        if existing_state:
            print("🔄 Resuming from saved state...")
            stack = existing_state["stack"]
            current_task = existing_state["current_task"]
            results = existing_state["results"]
            self.stats = existing_state.get("stats", self.stats)
        else:
            # Initialize new analysis
            print("🚀 Starting new RLM analysis...")
            stack = []
            current_task = Task(
                agent_type="Explorer",
                task_description=f"Analyze document at {document_path} for: {query}",
                context={
                    "document_path": document_path,
                    "query": query,
                    "analysis_start": datetime.now().isoformat()
                },
                depth=0
            )
            results = {}

        # Trampoline loop
        iteration = 0
        while True:
            iteration += 1

            # Safety check
            if iteration > 1000:
                raise RuntimeError("Exceeded maximum iterations (possible infinite loop)")

            if current_task.depth > self.config["max_recursion_depth"]:
                raise RuntimeError(f"Exceeded max recursion depth: {self.config['max_recursion_depth']}")

            # Execute current task
            result = self.dispatch_subagent(current_task)

            # Update cost tracking
            if isinstance(result, dict) and "cost_usd" in result:
                self.stats["total_cost_usd"] += result["cost_usd"]
                self.stats["total_tokens"] += result.get("token_count", 0)

            # Check if subagent wants to recurse
            if self.is_continuation(result):
                print(f"🔄 Subagent requests continuation: {result['agent_type']}")

                # Push current task to stack
                stack.append(current_task)

                # Create new task from continuation
                current_task = self.create_task_from_continuation(current_task, result)

                # Save state
                self.save_state(stack, current_task, results)

                print(f"📊 Stack depth: {len(stack)}, Current depth: {current_task.depth}")
                continue  # Trampoline!

            # Result completed - do we have a parent waiting?
            if stack:
                print(f"✅ Subagent completed, returning to parent")

                # Store result
                if current_task.return_to:
                    results[current_task.return_to] = result

                # Cache the result
                if isinstance(result, AnalysisResult):
                    self.save_cache(current_task, result)

                # Pop parent task
                parent_task = stack.pop()

                # Inject child results
                parent_task.child_results = results.copy()

                current_task = parent_task

                # Save state
                self.save_state(stack, current_task, results)

                print(f"🔙 Resumed parent at depth {current_task.depth}")
                continue

            # Stack empty - we're done!
            print(f"\n🎉 Analysis complete!")
            print(f"📊 Stats:")
            print(f"   Total subagent calls: {self.stats['total_subagent_calls']}")
            print(f"   Max depth reached: {self.stats['max_depth_reached']}")
            print(f"   Cache hits: {self.stats['cache_hits']}")
            print(f"   Total tokens: {self.stats['total_tokens']:,}")
            print(f"   Total cost: ${self.stats['total_cost_usd']:.4f}")

            # Clear state
            self.clear_state()

            return result

    def get_status(self) -> Dict:
        """Get current analysis status"""
        state = self.load_state()

        if not state:
            return {"status": "idle", "message": "No active analysis"}

        return {
            "status": "active",
            "current_depth": state["current_task"].depth,
            "stack_size": len(state["stack"]),
            "results_count": len(state["results"]),
            "stats": state["stats"]
        }

# CLI interface for testing
if __name__ == "__main__":
    import sys

    if len(sys.argv) < 2:
        print("Usage: python orchestrator.py <document_path> <query>")
        print("   or: python orchestrator.py status")
        sys.exit(1)

    orchestrator = RLMOrchestrator()

    if sys.argv[1] == "status":
        status = orchestrator.get_status()
        print(json.dumps(status, indent=2))
    else:
        document_path = sys.argv[1]
        query = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else "Analyze this document"

        result = orchestrator.analyze_document(document_path, query)
        print("\n" + "="*80)
        print("FINAL RESULT")
        print("="*80)
        print(json.dumps(result, indent=2))
